Wrong SQLALCHEMY_DATABASE_URI for DB_VENDOR mariadb or mysql

PrivacyIDEA does not start with DB_VENDOR set to mariadb or mysql

Traceback:

Traceback (most recent call last):
  File "/opt/privacyidea/bin/pi-manage", line 1739, in <module>
    manager.run()
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_script/__init__.py", line 417, in run
    result = self.handle(argv[0], argv[1:])
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_script/__init__.py", line 386, in handle
    res = handle(*args, **config)
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_script/commands.py", line 216, in __call__
    return self.run(*args, **kwargs)
  File "/opt/privacyidea/bin/pi-manage", line 589, in create_tables
    print(db)
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 1115, in __repr__
    self.engine.url if self.app or current_app else None
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 998, in engine
    return self.get_engine()
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 1017, in get_engine
    return connector.get_engine()
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 594, in get_engine
    self._engine = rv = self._sa.create_engine(sa_url, options)
  File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 1027, in create_engine
    return sqlalchemy.create_engine(sa_url, **engine_opts)
  File "<string>", line 2, in create_engine
  File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 375, in warned
    return fn(*args, **kwargs)
  File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 518, in create_engine
    entrypoint = u._get_entrypoint()
  File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/engine/url.py", line 662, in _get_entrypoint
    cls = registry.load(name)
  File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 343, in load
    raise exc.NoSuchModuleError(
sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:pymysql

Reason:

SQLALCHEMY_DATABASE_URI in /etc/privacyidea/pi.cfg should start with mysql+pymysql:// for DB_VENDOR mysql and mariadb+pymysql:// for DB_VENDOR mariadb. The current SQLALCHEMY_DATABASE_URI starting only with pymysql:// causes the error.

I am using v3.9.0 with DB_VENDOR=mariadb.

I resolved the problem with a quick patch in /mnt/privacyidea/scripts/, because I use the image in production at work.

--- /etc/privacyidea/pi.cfg
+++ /etc/privacyidea/pi.cfg
@@ -13,7 +13,7 @@
     sys.exit(1)
 # The realm, where users are allowed to login as administrators
 SUPERUSER_REALM = os.environ.get('SUPERUSER_REALM', ['administrator'])
-SQLALCHEMY_DATABASE_URI = "pymysql://pi:#####@mariadb:3306/pi"
+SQLALCHEMY_DATABASE_URI = "mariadb+pymysql://pi:#####@mariadb:3306/pi"
 PI_ENCFILE = os.environ.get("PI_ENCFILE", "/data/privacyidea/keys/encfile")
 PI_HSM = os.environ.get("PI_HSM", "default")
 PI_AUDIT_NO_SIGN = os.environ.get("PI_AUDIT_NO_SIGN", "False").lower() == "true"

Normally, this should be solved when building the URI in /usr/local/bin/start_privacyidea.sh. Maybe this should work:

...
    # Check the selected database vendor
    if [ "${DB_VENDOR}" = "mariadb" ]; then
        echo "[INFO] Using $DB_VENDOR ..."

        # Ensure that the necessary variables are defined
        [ -z "$DB_HOST" ] && echo "[ERROR] DB_HOST should be defined" && return 1
        [ -z "$DB_USER" ] && echo "[ERROR] DB_USER should be defined" && return 1
        [ -z "$DB_PASSWORD" ] && echo "[ERROR] DB_PASSWORD should be defined" && return 1
        [ -z "$DB_NAME" ] && echo "[ERROR] DB_NAME should be defined" && return 1

        # Set the default port if it is not defined
        if [ -z "$DB_PORT" ]; then
            echo DB_PORT is not defined using default port
            export DB_PORT=3306
        fi

        # Define the SQLAlchemy database URI using the necessary variables
        export SQLALCHEMY_DATABASE_URI=mariadb+pymysql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}

    elif [ "${DB_VENDOR}" = "mysql" ]; then
        echo "[INFO] Using $DB_VENDOR ..."

        # Ensure that the necessary variables are defined
        [ -z "$DB_HOST" ] && echo "[ERROR] DB_HOST should be defined" && return 1
        [ -z "$DB_USER" ] && echo "[ERROR] DB_USER should be defined" && return 1
        [ -z "$DB_PASSWORD" ] && echo "[ERROR] DB_PASSWORD should be defined" && return 1
        [ -z "$DB_NAME" ] && echo "[ERROR] DB_NAME should be defined" && return 1

        # Set the default port if it is not defined
        if [ -z "$DB_PORT" ]; then
            echo DB_PORT is not defined using default port
            export DB_PORT=3306
        fi

        # Define the SQLAlchemy database URI using the necessary variables
        export SQLALCHEMY_DATABASE_URI=mysql+pymysql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}

    elif [ "${DB_VENDOR}" = "postgresql" ]; then
...

Without the patch, I get this error:


2023-09-22 09:43:45 [INFO] Generating privacyIDEA database tables...
2023-09-22 09:43:45 
2023-09-22 09:43:46 Can not import grpc modules.
2023-09-22 09:43:47 
2023-09-22 09:43:47              _                    _______  _______
2023-09-22 09:43:47    ___  ____(_)  _____ _______ __/  _/ _ \/ __/ _ |
2023-09-22 09:43:47   / _ \/ __/ / |/ / _ `/ __/ // // // // / _// __ |
2023-09-22 09:43:47  / .__/_/ /_/|___/\_,_/\__/\_, /___/____/___/_/ |_|
2023-09-22 09:43:47 /_/                       /___/
2023-09-22 09:43:47                                                v3.9
2023-09-22 09:43:47     
2023-09-22 09:43:47 Traceback (most recent call last):
2023-09-22 09:43:47   File "/opt/privacyidea/bin/pi-manage", line 1739, in <module>
2023-09-22 09:43:47     manager.run()
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_script/__init__.py", line 417, in run
2023-09-22 09:43:47     result = self.handle(argv[0], argv[1:])
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_script/__init__.py", line 386, in handle
2023-09-22 09:43:47     res = handle(*args, **config)
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_script/commands.py", line 216, in __call__
2023-09-22 09:43:47     return self.run(*args, **kwargs)
2023-09-22 09:43:47   File "/opt/privacyidea/bin/pi-manage", line 589, in create_tables
2023-09-22 09:43:47     print(db)
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 1115, in __repr__
2023-09-22 09:43:47     self.engine.url if self.app or current_app else None
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 998, in engine
2023-09-22 09:43:47     return self.get_engine()
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 1017, in get_engine
2023-09-22 09:43:47     return connector.get_engine()
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 594, in get_engine
2023-09-22 09:43:47     self._engine = rv = self._sa.create_engine(sa_url, options)
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/flask_sqlalchemy/__init__.py", line 1027, in create_engine
2023-09-22 09:43:47     return sqlalchemy.create_engine(sa_url, **engine_opts)
2023-09-22 09:43:47   File "<string>", line 2, in create_engine
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 375, in warned
2023-09-22 09:43:47     return fn(*args, **kwargs)
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 518, in create_engine
2023-09-22 09:43:47     entrypoint = u._get_entrypoint()
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/engine/url.py", line 662, in _get_entrypoint
2023-09-22 09:43:47     cls = registry.load(name)
2023-09-22 09:43:47   File "/opt/privacyidea/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 343, in load
2023-09-22 09:43:47     raise exc.NoSuchModuleError(
2023-09-22 09:43:47 sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:pymysql

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -35,7 +35,9 @@ jobs:
             type=sha
             # set latest tag for default branch
             type=raw,value=latest,enable={{is_default_branch}}
-            type=ref,event=tag
+            type=semver,pattern=v{{version}}
+            type=semver,pattern=v{{major}}.{{minor}}
+
       - name: Set up QEMU
         uses: docker/setup-qemu-action@v3
       -

--- a/rootfs/usr/local/bin/start_privacyidea.sh
+++ b/rootfs/usr/local/bin/start_privacyidea.sh
@@ -22,51 +22,56 @@ function main {
 function generate_pi_config {
 
     # Check the selected database vendor
-    if [ "${DB_VENDOR}" = "mariadb" ] || [ "${DB_VENDOR}" = "mysql" ]; then
-        echo "[INFO] Using $DB_VENDOR ..."
+    case $DB_VENDOR in
+        "mariadb" | "mysql")
+            echo "[INFO] Using $DB_VENDOR ..."
 
-        # Ensure that the necessary variables are defined
-        [ -z "$DB_HOST" ] && echo "[ERROR] DB_HOST should be defined" && return 1
-        [ -z "$DB_USER" ] && echo "[ERROR] DB_USER should be defined" && return 1
-        [ -z "$DB_PASSWORD" ] && echo "[ERROR] DB_PASSWORD should be defined" && return 1
-        [ -z "$DB_NAME" ] && echo "[ERROR] DB_NAME should be defined" && return 1
+            # Ensure that the necessary variables are defined
+            [ -z "$DB_HOST" ] && echo "[ERROR] DB_HOST should be defined" && return 1
+            [ -z "$DB_USER" ] && echo "[ERROR] DB_USER should be defined" && return 1
+            [ -z "$DB_PASSWORD" ] && echo "[ERROR] DB_PASSWORD should be defined" && return 1
+            [ -z "$DB_NAME" ] && echo "[ERROR] DB_NAME should be defined" && return 1
 
-        # Set the default port if it is not defined
-        if [ -z "$DB_PORT" ]; then
-            echo DB_PORT is not defined using default port
-            export DB_PORT=3306
-        fi
+            # Set the default port if it is not defined
+            if [ -z "$DB_PORT" ]; then
+                echo DB_PORT is not defined using default port
+                export DB_PORT=3306
+            fi
 
-        # Define the SQLAlchemy database URI using the necessary variables
-        export SQLALCHEMY_DATABASE_URI=pymysql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}
+            # Define the SQLAlchemy database URI using the necessary variables
+            export SQLALCHEMY_DATABASE_URI=${DB_VENDOR}+pymysql://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}
+            echo ${SQLALCHEMY_DATABASE_URI}
+            ;;
 
-    elif [ "${DB_VENDOR}" = "postgresql" ]; then
-        echo "Using $DB_VENDOR..."
+        "postgresql")
+            echo "Using $DB_VENDOR..."
 
-        # Ensure that the necessary variables are defined
-        [ -z "$DB_HOST" ] && echo "[ERROR] DB_HOST should be defined" && return 1
-        [ -z "$DB_USER" ] && echo "[ERROR] DB_USER should be defined" && return 1
-        [ -z "$DB_PASSWORD" ] && echo "[ERROR] DB_PASSWORD should be defined" && return 1
-        [ -z "$DB_NAME" ] && echo "[ERROR] DB_NAME should be defined" && return 1
+            # Ensure that the necessary variables are defined
+            [ -z "$DB_HOST" ] && echo "[ERROR] DB_HOST should be defined" && return 1
+            [ -z "$DB_USER" ] && echo "[ERROR] DB_USER should be defined" && return 1
+            [ -z "$DB_PASSWORD" ] && echo "[ERROR] DB_PASSWORD should be defined" && return 1
+            [ -z "$DB_NAME" ] && echo "[ERROR] DB_NAME should be defined" && return 1
 
-        # Set the default port if it is not defined
-        if [ -z "$DB_PORT" ]; then
-            echo "[INFO] DB_PORT is not defined using default port 5432"
+            # Set the default port if it is not defined
+            if [ -z "$DB_PORT" ]; then
+                echo "[INFO] DB_PORT is not defined using default port 5432"
+                echo ""
+                export DB_PORT=5432
+            fi
+
+            # Define the SQLAlchemy database URI using the necessary variables
+            export SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}
+            ;;
+
+        *)
             echo ""
-            export DB_PORT=5432
-        fi
+            echo "[WARNING] DB_VENDOR environment variable is not set. Using default SQLite..."
+            echo ""
 
-        # Define the SQLAlchemy database URI using the necessary variables
-        export SQLALCHEMY_DATABASE_URI=postgresql+psycopg2://${DB_USER}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_NAME}
-
-    else
-        echo ""
-        echo "[WARNING] DB_VENDOR environment variable is not set. Using default SQLite..."
-        echo ""
-
-        # Define the SQLAlchemy database URI for SQLite
-        export SQLALCHEMY_DATABASE_URI=sqlite://///data/privacyidea/privacyidea.db
-    fi
+            # Define the SQLAlchemy database URI for SQLite
+            export SQLALCHEMY_DATABASE_URI=sqlite://///data/privacyidea/privacyidea.db
+            ;;
+    esac
 
     # Check if the configuration file already exists
     if [ ! -f /etc/privacyidea/pi.cfg ]; then

